auditlog/proxy_decorator_test.go

```diff
@@ -0,0 +1,159 @@
+package auditlog
+
+import (
+	"bytes"
+	"io/ioutil"
+	"net/http"
+	"regexp"
+	"sync"
+	"testing"
+
+	"github.com/ory/x/logrusx"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/mock"
+)
+
+type MockProxy struct {
+	mock.Mock
+	t         *testing.T
+	roundTrip func(t *testing.T, r *http.Request) (*http.Response, error)
+	director  func(t *testing.T, r *http.Request)
+}
+
+func (m *MockProxy) RoundTrip(r *http.Request) (*http.Response, error) {
+	m.Called(r)
+
+	return m.roundTrip(m.t, r)
+}
+
+func (m *MockProxy) Director(r *http.Request) {
+	m.Called(r)
+	m.director(m.t, r)
+}
+
+type MockSender struct {
+	mock.Mock
+	mx *sync.Mutex
+}
+
+func (m *MockSender) Send(e Event) {
+	m.Called(e)
+
+	// Unlock mutex to run tests in parent goroutine.
+	m.mx.Unlock()
+}
+
+func TestProxyAuditLogDecorator_Director(t *testing.T) {
+	request := &http.Request{}
+	proxy := &MockProxy{
+		director: func(t *testing.T, r *http.Request) {},
+	}
+
+	decorator := ProxyAuditLogDecorator{proxy: proxy}
+
+	proxy.On("Director", request).Return()
+	decorator.Director(request)
+	proxy.AssertExpectations(t)
+}
+
+func TestProxyAuditLogDecorator_RoundTrip(t *testing.T) {
+	tests := []struct {
+		request  *http.Request
+		response *http.Response
+		hasErr   bool
+		proxy    *MockProxy
+		builders []EventBuilder
+		senders  []Sender
+	}{
+		{
+			request:  &http.Request{},
+			response: &http.Response{},
+			proxy: &MockProxy{
+				roundTrip: func(t *testing.T, r *http.Request) (*http.Response, error) {
+					return &http.Response{}, nil
+				},
+			},
+			builders: []EventBuilder{},
+			senders:  []Sender{},
+		},
+		{
+			request:  &http.Request{},
+			response: &http.Response{Body: ioutil.NopCloser(bytes.NewReader([]byte("test")))},
+			proxy: &MockProxy{
+				roundTrip: func(t *testing.T, r *http.Request) (*http.Response, error) {
+					return &http.Response{Body: ioutil.NopCloser(bytes.NewReader([]byte("test")))}, nil
+				},
+			},
+			builders: []EventBuilder{},
+			senders:  []Sender{},
+		},
+	}
+
+	for _, tst := range tests {
+		decorator := ProxyAuditLogDecorator{proxy: tst.proxy}
+
+		tst.proxy.On("RoundTrip", tst.request).Return(tst.response, nil)
+		resp, err := decorator.RoundTrip(tst.request)
+
+		assert.Equal(t, resp, tst.response)
+		assert.Nil(t, err)
+		tst.proxy.AssertExpectations(t)
+	}
+}
+
+func TestProxyAuditLogDecorator_RoundTrip2(t *testing.T) {
+	request, _ := http.NewRequest("GET", "http://localhost:8080/return200", nil)
+	response := &http.Response{Body: ioutil.NopCloser(bytes.NewReader([]byte("test")))}
+	proxy := &MockProxy{
+		roundTrip: func(t *testing.T, r *http.Request) (*http.Response, error) {
+			return &http.Response{Body: ioutil.NopCloser(bytes.NewReader([]byte("test")))}, nil
```

Comment:
Can you explain to me why not use the `response` defined earlier instead of `&http.Response{Body: ioutil.NopCloser(bytes.NewReader([]byte("test")))}`? Is it not thread-safe or I just dont get the test? :)